### PR TITLE
Abstraxion impl core

### PR DESCRIPTION
**Pull request by BurntVal**
Reopen of the abstraxion implementing core PR

CONTEXT: Existing PR was far out of sync. This PR aims to cleanup the `AbstraxionAuth` class from `abstraxion-core` and implement the core package in `abstraxion` to keep code uniformity and reduce redundancy

### CODE REVIEW

File: packages/abstraxion/src/components/AbstraxionSignin/index.tsx [hunk 1]
```diff
@@ -1,177 +1,31 @@
...
-          // If there isn't an existing keypair, or there isn't a granter in either localStorage or the url params, we want to start from scratch
-          // Generate new keypair and redirect to dashboard
-          const newKeypairaddress = await generateAndStoreTempAccount(); // remove and replace existing keypair
-          const { dashboardUrl } = await fetchConfig(rpcUrl);
-          openDashboardTab(newKeypairaddress, contracts, dashboardUrl);
+        if (!abstraxionAuth) {
+          throw new Error("abstraxion-core not initialized");
         }
+        setIsConnecting(true);
+        await abstraxionAuth.login();
```
Comment (justinbarry) [inline packages/abstraxion/src/components/AbstraxionSignin/index.tsx:26]: What happens if this is called multiple times?

File: packages/abstraxion-core/src/AbstraxionAuth.ts [hunk 2]
```diff
@@ -248,98 +296,105 @@ export class AbstraxionAuth {
...
+  /**
+   * Initiates the login process for the user.
+   * Checks if a local keypair and granter address exist, either from URL parameters or localStorage.
+   * If both exist, polls for grants and updates the authentication state if successful.
+   * If not, generates a new keypair and redirects to the dashboard for grant issuance.
+   *
+   * @returns {Promise<void>} - A Promise that resolves once the login process is complete.
+   * @throws {Error} - If the login process encounters an error.
+   */
+  async login(): Promise<void> {
```
Comment (justinbarry) [inline packages/abstraxion-core/src/AbstraxionAuth.ts:350]: This method needs to be resilient to being called several times while the login process is in progress.

**Review by justinbarry — CHANGES_REQUESTED** [review]
Love the direction! Left some comments.

File: packages/abstraxion/src/components/AbstraxionSignin/index.tsx [hunk 3]
```diff
@@ -1,177 +1,31 @@
...
-        const delay = Math.pow(2, retries) * 1000;
-        await new Promise((resolve) => setTimeout(resolve, delay));
-        retries++;
-      }
-    }
-
-    console.error("Max retries exceeded, giving up.");
-    return false;
-  }
+  const retryRedirect = async () => {
```
[the remark this answers is not in the document]
Comment (BurntVal) [reply packages/abstraxion/src/components/AbstraxionSignin/index.tsx:14]: The actual redirect does (and it's all contained within the login fn), this particular function is just for the designed modal screen we have that allows users to explicitly redirect if for some reason it hasn't happened automatically for them

File: packages/abstraxion/src/components/Abstraxion/index.tsx [hunk 4]
```diff
@@ -7,19 +7,21 @@ import {
   ContractGrantDescription,
   SpendLimit,
 } from "../AbstraxionContext";
-import { Loading } from "../Loading";
 import { ErrorDisplay } from "../ErrorDisplay";
 import { AbstraxionSignin } from "../AbstraxionSignin";
```
[the remark this answers is not in the document]
Comment (BurntVal) [reply packages/abstraxion/src/components/Abstraxion/index.tsx:11]: What part are you referring to exactly

File: packages/abstraxion-core/src/AbstraxionAuth.ts [hunk 5]
```diff
@@ -173,11 +198,6 @@ export class AbstraxionAuth {
         },
       );
 
-      const wallet = await this.getLocalKeypair();
-      if (wallet) {
-        this.signArbWallet = wallet;
-      }
-
       this.client = directClient;
```
[the remark this answers is not in the document]
Comment (BurntVal) [reply packages/abstraxion-core/src/AbstraxionAuth.ts:201]: We do, and on line 164, if an instance already exists it just returns it

File: packages/abstraxion-core/src/AbstraxionAuth.ts [hunk 6]
```diff
@@ -187,8 +207,28 @@ export class AbstraxionAuth {
     }
   }
 
-  // Redirects to dashboard in order to issue claim with XION meta account for local keypair
-  openDashboardTab(userAddress: string): void {
+  /**
+   * Get dashboard url and redirect in order to issue claim with XION meta account for local keypair.
+   */
+  async redirectToDashboard(userAddress: string) {
```
[the remark this answers is not in the document]
Comment (BurntVal) [reply packages/abstraxion-core/src/AbstraxionAuth.ts:213]: I feel like I had a particular reason for this but it's not coming to mind. I can have this be self contained and see if maybe the reasoning strikes me back or if not it'll just be cleaner

File: packages/abstraxion-core/src/AbstraxionAuth.ts [hunk 7]
```diff
@@ -187,8 +207,28 @@ export class AbstraxionAuth {
...
+        "Something went wrong trying to redirect to XION dashboard: ",
+        error,
+      );
+    }
+  }
+
+  /**
+   * Configure URL and redirect page
+   */
+  private openDashboardTab(dashboardUrl: string, userAddress: string): void {
```
[the remark this answers is not in the document]
Comment (BurntVal) [reply packages/abstraxion-core/src/AbstraxionAuth.ts:231]: Will do

File: packages/abstraxion/src/components/AbstraxionSignin/index.tsx (hunk 1 above)
Comment (BurntVal) [reply packages/abstraxion/src/components/AbstraxionSignin/index.tsx:26]: Well depends on user env/state, but I definitely see your point. See reply to comment below

File: packages/abstraxion-core/src/AbstraxionAuth.ts (hunk 2 above)
Comment (BurntVal) [reply packages/abstraxion-core/src/AbstraxionAuth.ts:350]: Definitely agree. Maybe some instance state that a login has already been triggered. Any ideas on actions? Just simply return? Maybe a warning that a login is already in progress, etc?

**Comment by justinbarry** [comment]
image

@BurntVal this is showing on return to the demo.

File: packages/abstraxion/src/components/Abstraxion/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (justinbarry) [reply packages/abstraxion/src/components/Abstraxion/index.tsx:11]: `import { AbstraxionSignin } from "../AbstraxionSignin";` -> `import { AbstraxionContext } from "../AbstraxionContext";` -> `import { abstraxionAuth } from "../Abstraxion";`

File: packages/abstraxion-core/src/AbstraxionAuth.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (justinbarry) [reply packages/abstraxion-core/src/AbstraxionAuth.ts:202]: You can have a potential race condition. Relevant code:
```typescript
    if (this.client) {
      return this.client;
    }
    if (!this.rpcUrl) {
      throw new Error("Configuration not initialized");
    }
    if (!this.abstractAccount) {
      throw new Error("No account found.");
    }
    const granterAddress = this.getGranter();
    if (!granterAddress) {
      throw new Error("No granter found.");
    }
    const granteeAddress = await this.abstractAccount
      .getAccounts()
      .then((accounts: any) => {
        if (accounts.length === 0) {
          throw new Error("No account found.");
        }
        return accounts[0].address;
      });
    const directClient = await GranteeSignerClient.connectWithSigner(
      this.rpcUrl,
      this.abstractAccount,
      {
        gasPrice: GasPrice.fromString("0uxion"),
        granterAddress,
        granteeAddress,
      },
    );

    this.client = directClient;
```

1.`getSigner` is called
2. 'getSigner executes until it hits `const granteeAddress = await this.abstractAccount` and waits for the async process to complete
3. `getSigner` is called again for whatever reason
4. Invocation 2 is now paused at `const granteeAddress = await this.abstractAccount`
5. Invocation 1 now proceeds (Just assume the following promises resolve in FIFO order) and sets `this.client`
6. Invocation 2 now proceeds (Just assume the following promises resolve in FIFO order) and sets `this.client`
7. The object returned by invocation 1 will not be the same object as returned by the second (Although the functionality would be the same)

Just something to be aware of if we are using strict equality to trigger another process (And to reduce the RPCs calls slightly).

**Review by justinbarry — APPROVED** [review]

**Pull request merged**